Access to the tycho .cache directory is not properly synchronized

I am just upgrading from tycho 2.2.0 to 2.6.0 and now my CI build (using maven parallel builds with `-T 1C`) sporadically fails with this error during dependency resolution:
```
[INFO] Computing target platform for MavenProject: com.foo:bar:1.0.0-SNAPSHOT @ ...
[ERROR] Internal error: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: IOException while extracting 'stax2-api-3.1.4.jar' from /opt/jenkins/workspace/MYPIPELINE/.repository/com/example/somedependency/1.2.3/somedependency-1.2.3.jar: /opt/jenkins/workspace/MYPIPELINE/.repository/.cache/tycho/somedependency-1.2.3.jar/stax2-api-3.1.4.jar -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: IOException while extracting 'stax2-api-3.1.4.jar' from /opt/jenkins/workspace/MYPIPELINE/.repository/com/example/somedependency/1.2.3/somedependency-1.2.3.jar
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:120)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: IOException while extracting 'stax2-api-3.1.4.jar' from /opt/jenkins/workspace/MYPIPELINE/.repository/com/example/somedependency/1.2.3/somedependency-1.2.3.jar
    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0 (Native Method)
    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance (NativeConstructorAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance (DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance (Constructor.java:490)
    at java.util.concurrent.ForkJoinTask.getThrowableException (ForkJoinTask.java:600)
    at java.util.concurrent.ForkJoinTask.get (ForkJoinTask.java:1006)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.resolveProjects (TychoMavenLifecycleParticipant.java:164)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:105)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:264)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

...

Caused by: java.nio.file.FileAlreadyExistsException: /opt/jenkins/workspace/MYPIPELINE/.repository/.cache/tycho/somedependency-1.2.3.jar/stax2-api-3.1.4.jar
    at sun.nio.fs.UnixException.translateToIOException (UnixException.java:94)
    at sun.nio.fs.UnixException.rethrowAsIOException (UnixException.java:111)
    at sun.nio.fs.UnixException.rethrowAsIOException (UnixException.java:116)
    at sun.nio.fs.UnixFileSystemProvider.newByteChannel (UnixFileSystemProvider.java:215)
    at java.nio.file.spi.FileSystemProvider.newOutputStream (FileSystemProvider.java:478)
    at java.nio.file.Files.newOutputStream (Files.java:219)
    at java.nio.file.Files.copy (Files.java:3066)
    at org.eclipse.tycho.core.osgitools.DefaultBundleReader.copyStreamToFile (DefaultBundleReader.java:172)
    at org.eclipse.tycho.core.osgitools.DefaultBundleReader.extractZipEntries (DefaultBundleReader.java:151)
    at org.eclipse.tycho.core.osgitools.DefaultBundleReader.getEntry (DefaultBundleReader.java:126)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getNestedJarOrDir (OsgiBundleProject.java:547)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getBundleClasspath (OsgiBundleProject.java:513)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.resolveClassPath (OsgiBundleProject.java:231)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:173)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.lambda$resolveProjects$0 (TychoMavenLifecycleParticipant.java:147)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:183)
    at java.util.stream.SpinedBuffer.forEach (SpinedBuffer.java:250)
    at java.util.stream.Nodes$SpinedNodeBuilder.forEach (Nodes.java:1270)
    at java.util.stream.Nodes$InternalNodeSpliterator$OfRef.forEachRemaining (Nodes.java:1105)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:484)
    at java.util.stream.ForEachOps$ForEachTask.compute (ForEachOps.java:290)
    at java.util.concurrent.CountedCompleter.exec (CountedCompleter.java:746)
    at java.util.concurrent.ForkJoinTask.doExec (ForkJoinTask.java:290)
    at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec (ForkJoinPool.java:1020)
    at java.util.concurrent.ForkJoinPool.scan (ForkJoinPool.java:1656)
    at java.util.concurrent.ForkJoinPool.runWorker (ForkJoinPool.java:1594)
    at java.util.concurrent.ForkJoinWorkerThread.run (ForkJoinWorkerThread.java:177)
```

`somedependency-1.2.3.jar` is a proprietary lib which includes a bunch of other third party jars.

It looks like a concurrency issue somewhere in the dependency resolution when using parallel builds and multiple plugins have the same dependency but in a different target platform?

The interesting codepath is this with nested jars of the same name:

```
Caused by: java.nio.file.FileAlreadyExistsException: /opt/jenkins/workspace/MYPIPELINE/.repository/.cache/tycho/somedependency-1.2.3.jar/stax2-api-3.1.4.jar
    at sun.nio.fs.UnixException.translateToIOException (UnixException.java:94)
    at sun.nio.fs.UnixException.rethrowAsIOException (UnixException.java:111)
    at sun.nio.fs.UnixException.rethrowAsIOException (UnixException.java:116)
    at sun.nio.fs.UnixFileSystemProvider.newByteChannel (UnixFileSystemProvider.java:215)
    at java.nio.file.spi.FileSystemProvider.newOutputStream (FileSystemProvider.java:478)
    at java.nio.file.Files.newOutputStream (Files.java:219)
    at java.nio.file.Files.copy (Files.java:3066)
    at org.eclipse.tycho.core.osgitools.DefaultBundleReader.copyStreamToFile (DefaultBundleReader.java:172)
    at org.eclipse.tycho.core.osgitools.DefaultBundleReader.extractZipEntries (DefaultBundleReader.java:151)
    at org.eclipse.tycho.core.osgitools.DefaultBundleReader.getEntry (DefaultBundleReader.java:126)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getNestedJarOrDir (OsgiBundleProject.java:547)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getBundleClasspath (OsgiBundleProject.java:513)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.resolveClassPath (OsgiBundleProject.java:231)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:173)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.lambda$resolveProjects$0 (TychoMavenLifecycleParticipant.java:147)
```

--- a/tycho-core/src/main/java/org/eclipse/tycho/core/osgitools/DefaultBundleReader.java
+++ b/tycho-core/src/main/java/org/eclipse/tycho/core/osgitools/DefaultBundleReader.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2008, 2011 Sonatype Inc. and others.
+ * Copyright (c) 2008, 2022 Sonatype Inc. and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -7,6 +7,7 @@
  *
  * Contributors:
  *    Sonatype Inc. - initial API and implementation
+ *    Christoph Läubrich - Issue #663 - Access to the tycho .cache directory is not properly synchronized 
  *******************************************************************************/
 package org.eclipse.tycho.core.osgitools;
 
@@ -18,9 +19,10 @@ import java.nio.file.Files;
 import java.nio.file.StandardCopyOption;
 import java.util.Enumeration;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.Map;
-import java.util.Set;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.jar.JarFile;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
@@ -34,11 +36,12 @@ import org.eclipse.tycho.locking.facade.FileLocker;
 @Component(role = BundleReader.class)
 public class DefaultBundleReader extends AbstractLogEnabled implements BundleReader {
 
+    private static final long LOCK_TIMEOUT = Long.getLong("tycho.bundlereader.lock.timeout", 5 * 60 * 1000L);
     public static final String CACHE_PATH = ".cache/tycho";
     private final Map<String, OsgiManifest> manifestCache = new HashMap<>();
 
     private File cacheDir;
-    private Set<String> extractedFiles = new HashSet<>();
+    private ConcurrentMap<String, Optional<File>> extractedFiles = new ConcurrentHashMap<>();
 
     @Requirement
     private FileLockService fileLockService;
@@ -109,32 +112,42 @@ public class DefaultBundleReader extends AbstractLogEnabled implements BundleRea
             getLogger().warn("Ignoring Bundle-ClassPath entry '" + path + "' of bundle " + bundleLocation);
             return null;
         }
-        final File result;
+        final Optional<File> result;
         if (bundleLocation.isDirectory()) {
-            result = new File(bundleLocation, path);
-        } else {
-            try {
-                File outputDirectory = new File(cacheDir, bundleLocation.getName());
-                result = new File(outputDirectory, path);
-                String resultPath = result.getCanonicalPath();
-                if (extractedFiles.contains(resultPath) && result.exists()) {
-                    return result;
-                } else {
-                    FileLocker locker = fileLockService.getFileLocker(outputDirectory);
-                    locker.lock(5 * 60 * 1000L);
-                    try {
-                        extractZipEntries(bundleLocation, path, outputDirectory);
-                    } finally {
-                        locker.release();
-                    }
-                    extractedFiles.add(resultPath);
-                }
-            } catch (IOException e) {
-                throw new RuntimeException("IOException while extracting '" + path + "' from " + bundleLocation, e);
+            File file = new File(bundleLocation, path);
+            if (file.exists()) {
+                result = Optional.of(file);
+            } else {
+                result = Optional.empty();
             }
+        } else {
+            String cacheKey;
+            File outputDirectory = new File(cacheDir, bundleLocation.getName());
+            File cacheFile = new File(outputDirectory, path);
+            try {
+                cacheKey = cacheFile.getCanonicalPath();
+            } catch (IOException e) {
+                throw new RuntimeException("can't get canonical path for " + cacheFile, e);
+            }
+            result = extractedFiles.computeIfAbsent(cacheKey, nil -> {
+                FileLocker locker = fileLockService.getFileLocker(outputDirectory);
+                locker.lock(LOCK_TIMEOUT);
+                try {
+                    extractZipEntries(bundleLocation, path, outputDirectory);
+                    if (cacheFile.exists()) {
+                        return Optional.of(cacheFile);
+                    }
+                    return Optional.empty();
+                } catch (IOException e) {
+                    throw new RuntimeException(
+                            "Can't extract '" + path + "' from " + bundleLocation + " to " + outputDirectory, e);
+                } finally {
+                    locker.release();
+                }
+            });
         }
-        if (result.exists()) {
-            return result;
+        if (result.isPresent()) {
+            return result.get();
         } else {
             getLogger().warn("Bundle-ClassPath entry " + path + " does not exist in " + bundleLocation);
             return null;
@@ -147,16 +160,16 @@ public class DefaultBundleReader extends AbstractLogEnabled implements BundleRea
             InputStream singleEntryStream;
             if (singleEntry != null && !singleEntry.isDirectory()
                     && (singleEntryStream = zip.getInputStream(singleEntry)) != null) {
-                // fix for performance bug 367098: avoid loop if path is a single zip file entry
-                copyStreamToFile(singleEntryStream, new File(outputDirectory, singleEntry.getName()),
-                        singleEntry.getTime());
+                // extract an exact match
+                File outputFile = new File(outputDirectory, singleEntry.getName());
+                copyStreamToFile(singleEntryStream, outputFile, singleEntry.getTime());
             } else {
-                // loop over all entries and extract matching
+                // loop over all entries and extract matching (e.g. in case we wan't a directory entry extracted)
                 for (Enumeration<? extends ZipEntry> entries = zip.entries(); entries.hasMoreElements();) {
                     ZipEntry zipEntry = entries.nextElement();
                     if (!zipEntry.isDirectory() && zipEntry.getName().startsWith(path)) {
-                        copyStreamToFile(zip.getInputStream(zipEntry), new File(outputDirectory, zipEntry.getName()),
-                                zipEntry.getTime());
+                        File outputFile = new File(outputDirectory, zipEntry.getName());
+                        copyStreamToFile(zip.getInputStream(zipEntry), outputFile, zipEntry.getTime());
                     }
                 }
             }

